csrc/optimization/layout_inference.cpp

```diff
@@ -0,0 +1,236 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/all_nodes.h>
+#include <ir/utils.h>
+#include <optimization/layout_inference.h>
+
+namespace nvfuser {
+
+namespace {
+
+class MemoryFormatInferencer : public OptOutConstDispatch {
+ public:
+  MemoryFormatInferencer(
+      std::unordered_map<const TensorView*, MemoryFormat>& format_map)
+      : format_map_(format_map) {}
+
+ private:
+  void handle(const UnaryOp*) override;
+  void handle(const BinaryOp*) override;
+  void handle(const BroadcastOp*) override;
+  // TODO: Add more propagation rules
+  //   void handle(const Reduction*) override;
+  //   void handle(const LoadStoreOp*) override;
+  //   void handle(const SqueezeOp*) override;
+  //   void handle(const ExpandOp*) override;
+
+ private:
+  // format_map_ records the stride order (memory format) of each TensorView.
+  // Since it only handles permutation from a rfactor domain to allocation
+  // domain, it can be interpreted as:
+  //
+  // e.g. TV0 rfactor domain [i0, i1, i2]
+  //            alloc domain [i0, i2, i1]
+  //           memory format   0,  2,  1
+  std::unordered_map<const TensorView*, MemoryFormat>& format_map_;
+};
+
+// UnaryOp propagation forward memory format from input to output
+void MemoryFormatInferencer::handle(const UnaryOp* op) {
+  TensorView* out = dynamic_cast<TensorView*>(op->out());
+  if (out == nullptr) {
+    return;
+  }
+  TensorView* in = op->in()->as<TensorView>();
+  if (auto iter = format_map_.find(in); iter != format_map_.end()) {
+    format_map_[out] = iter->second;
+  }
+}
+
+// BinaryOp propagation tries to merge the memory format of both inputs
+//
+//   1. when there's only one operand has a recorded memory format, it forwards
```

Comment:
For inputs without allocation domain, we're leaving them as empty, which sounds like a bad idea.

Meanwhile, this could still happen to tensors created with factory method. Since we are only recording memory format of input tensors and I don't want that to affect the output memory format.

This resonates with @jacobhinkle 's other comment on `should we have a backward propagation as well`.